In [2]:
# Import libraries
import pandas as pd
import numpy as np
import os

## Set path
path = r'/Users/leo/Desktop/Career Foundry/Achievement 4/Instacart Basket Analysis 30:10:25'

In [3]:
## Import Data
ords_prods_merge = pd.read_pickle(os.path.join(path,'02 - data','prepared data','ords_prods_merge_task_4.8_practice.pkl'))

# 2.In this Exercise, you learned how to find the aggregated mean of the “order_number” column grouped by “department_id” for a subset of your dataframe. Now, repeat this process for the entire dataframe.


In [4]:
## Entire data frame
ords_prods_merge.groupby('department_id').agg({'order_number': ['mean']})

order_number
                      mean
department_id             
1                15.457838
2                17.277920
3                17.170395
4                17.811403
5                15.215751
6                16.439806
7                17.225802
8                15.340650
9                15.895474
10               20.197148
11               16.170638
12               15.887671
13               16.583536
14               16.773669
15               16.165037
16               17.665606
17               15.694469
18               19.310397
19               17.177343
20               16.473447
21               22.902379

# 3. Analyze the result. How do the results for the entire dataframe differ from those of the subset? Include your comments in a markdown cell below the executed code.

When comparing the subset to the entire dataset, the mean order numbers by department are generally similar but slightly more stable across the full dataset. In the subset, some departments (like 10 and 21) showed higher averages, likely due to the smaller sample size amplifying variability. After including all data, these differences became less extreme - for example, department 21 dropped from 25.5 to 22.9, and department 10 increased slightly from 18.7 to 20.2.
Overall, the full dataset provides a more balanced and reliable representation of ordering behavior across all departments, while the subset results were more influenced by sample-specific trends.

# 4. Follow the instructions in the Exercise for creating a loyalty flag for existing customers using the transform() and loc() functions.

In [ ]:
# Aggregating Data with transform()
ords_prods_merge['max_order'] = ords_prods_merge.groupby(['user_id'])['order_number'].transform(np.max)

# Deriving Columns with loc()

In [ ]:
ords_prods_merge.loc[ords_prods_merge['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'

In [ ]:
ords_prods_merge.loc[(ords_prods_merge['max_order'] <= 40) & (ords_prods_merge['max_order'] > 10), 'loyalty_flag'] = 'Regular customer'

In [ ]:
ords_prods_merge.loc[ords_prods_merge['max_order'] <= 10, 'loyalty_flag'] = 'New customer'

In [ ]:
# Check frequency
ords_prods_merge['loyalty_flag'].value_counts(dropna=False)

In [5]:
ords_prods_merge.head(4)

product_id                product_name  aisle_id  department_id  prices  \
0           1  Chocolate Sandwich Cookies        61             19     5.8   
1           1  Chocolate Sandwich Cookies        61             19     5.8   
2           1  Chocolate Sandwich Cookies        61             19     5.8   
3           1  Chocolate Sandwich Cookies        61             19     5.8   

   order_id  user_id  order_number  order_dow  order_hour_of_day  \
0   3139998      138            28          6                 11   
1   1977647      138            30          6                 17   
2    389851      709             2          0                 21   
3    652770      764             1          3                 13   

   days_since_prior_order  First_order  add_to_cart_order  reordered _merge  \
0                     3.0        False                  5          0   both   
1                    20.0        False                  1          1   both   
2                     6.0        False                 20          0   both   
3                     NaN         True                 10          0   both   

     Busiest_days  Slowest_days busiest_period_of_day  max_order  \
0  Regularly busy                         Most orders         32   
1  Regularly busy                      Average orders         32   
2    Busiest days                       Fewest orders          5   
3  Regularly busy  Slowest days           Most orders          3   

       loyalty_flag  
0  Regular customer  
1  Regular customer  
2      New customer  
3      New customer

# 5.The marketing team at Instacart wants to know whether there’s a difference between the spending habits of the three types of customers you identified. Use the loyalty flag you created and check the basic statistics of the product prices for each loyalty category (Loyal Customer, Regular Customer, and New Customer). What you’re trying to determine is whether the prices of products purchased by loyal customers differ from those purchased by regular or new customers.

In [6]:
ords_prods_merge.groupby('loyalty_flag')['prices'].describe()

count       mean         std  min  25%  50%   75%  \
loyalty_flag                                                               
Loyal customer    10284093.0  10.386336  328.017787  1.0  4.2  7.4  11.2   
New customer       6243990.0  13.294670  597.560299  1.0  4.2  7.4  11.3   
Regular customer  15876776.0  12.495717  539.720919  1.0  4.2  7.4  11.3   

                      max  
loyalty_flag               
Loyal customer    99999.0  
New customer      99999.0  
Regular customer  99999.0

## Analysis on spending habits based on 3 type of customers.
The mean product prices differ slightly across the three loyalty categories.

New customers have the highest average product price (£13.29)

Regular customers spend slightly less on average (£12.50)

Loyal customers have the lowest average price (£10.39)

This suggests that new customers tend to purchase higher-priced products, while loyal customers may prefer lower-priced or more frequently purchased everyday items.
However, the large standard deviations (and extremely high maximum values) indicate the presence of outliers, so the overall differences should be interpreted with caution.

# 6.The team now wants to target different types of spenders in their marketing campaigns. This can be achieved by looking at the prices of the items people are buying. Create a spending flag for each user based on the average price across all their orders using the following criteria:
1. If the mean of the prices of products purchased by a user is lower than 10, then flag them as a “Low spender.”
2. If the mean of the prices of products purchased by a user is higher than or equal to 10, then flag them as a “High spender.”


In [7]:
## Calculate each user’s average price
ords_prods_merge['average_price'] = ords_prods_merge.groupby('user_id')['prices'].transform('mean')

In [8]:
## Create the spending_flag column
ords_prods_merge.loc[ords_prods_merge['average_price'] < 10, 'spending_flag'] = 'Low spender'

In [9]:
ords_prods_merge.loc[ords_prods_merge['average_price'] >= 10, 'spending_flag'] = 'High spender'

In [10]:
## Check results
ords_prods_merge['spending_flag'].value_counts()

spending_flag
Low spender     31770614
High spender      634245
Name: count, dtype: int64

In [11]:
## Check new columns
ords_prods_merge.head(4)

product_id                product_name  aisle_id  department_id  prices  \
0           1  Chocolate Sandwich Cookies        61             19     5.8   
1           1  Chocolate Sandwich Cookies        61             19     5.8   
2           1  Chocolate Sandwich Cookies        61             19     5.8   
3           1  Chocolate Sandwich Cookies        61             19     5.8   

   order_id  user_id  order_number  order_dow  order_hour_of_day  ...  \
0   3139998      138            28          6                 11  ...   
1   1977647      138            30          6                 17  ...   
2    389851      709             2          0                 21  ...   
3    652770      764             1          3                 13  ...   

   add_to_cart_order  reordered  _merge    Busiest_days  Slowest_days  \
0                  5          0    both  Regularly busy                 
1                  1          1    both  Regularly busy                 
2                 20          0    both    Busiest days                 
3                 10          0    both  Regularly busy  Slowest days   

  busiest_period_of_day max_order      loyalty_flag  average_price  \
0           Most orders        32  Regular customer       6.935811   
1        Average orders        32  Regular customer       6.935811   
2         Fewest orders         5      New customer       7.930208   
3           Most orders         3      New customer       4.972414   

  spending_flag  
0   Low spender  
1   Low spender  
2   Low spender  
3   Low spender  

[4 rows x 22 columns]

In [12]:
## Drop average_price as client does not need it
ords_prods_merge.drop(columns=['average_price'], inplace=True)

In [16]:
ords_prods_merge.head(2)

product_id                product_name  aisle_id  department_id  prices  \
0           1  Chocolate Sandwich Cookies        61             19     5.8   
1           1  Chocolate Sandwich Cookies        61             19     5.8   

   order_id  user_id  order_number  order_dow  order_hour_of_day  \
0   3139998      138            28          6                 11   
1   1977647      138            30          6                 17   

   days_since_prior_order  First_order  add_to_cart_order  reordered _merge  \
0                     3.0        False                  5          0   both   
1                    20.0        False                  1          1   both   

     Busiest_days Slowest_days busiest_period_of_day  max_order  \
0  Regularly busy                        Most orders         32   
1  Regularly busy                     Average orders         32   

       loyalty_flag spending_flag  
0  Regular customer   Low spender  
1  Regular customer   Low spender

# 7. In order to send relevant notifications to users within the app (for instance, asking users if they want to buy the same item again), the Instacart team wants you to determine frequent versus non-frequent customers. Create an order frequency flag that marks the regularity of a user’s ordering behavior according to the median in the “days_since_prior_order” column. The criteria for the flag should be as follows:
   
1. If the median of “days_since_prior_order” is higher than 20, then the customer should be labeled a “Non-frequent customer.”
2. If the median is higher than 10 and lower than or equal to 20, then the customer should be labeled a “Regular customer.”
3. If the median is lower than or equal to 10, then the customer should be labeled a “Frequent customer.”


In [15]:
pd.options.display.max_columns=None

In [17]:
## Calculate each users median number of days between orders
ords_prods_merge['median_days'] = ords_prods_merge.groupby('user_id')['days_since_prior_order'].transform('median')

In [18]:
## Create the order frequency flag
ords_prods_merge.loc[ords_prods_merge['median_days'] > 20, 'order_frequency_flag'] = 'Non-frequent customer'

In [19]:
ords_prods_merge.loc[(ords_prods_merge['median_days'] > 10) & (ords_prods_merge['median_days'] <= 20), 'order_frequency_flag'] = 'Regular customer'

In [20]:
ords_prods_merge.loc[ords_prods_merge['median_days'] <= 10, 'order_frequency_flag'] = 'Frequent customer'

In [21]:
## Check results
ords_prods_merge['order_frequency_flag'].value_counts()

order_frequency_flag
Frequent customer        21559853
Regular customer          7208564
Non-frequent customer     3636437
Name: count, dtype: int64

In [24]:
ords_prods_merge.head(3)

product_id                product_name  aisle_id  department_id  prices  \
0           1  Chocolate Sandwich Cookies        61             19     5.8   
1           1  Chocolate Sandwich Cookies        61             19     5.8   
2           1  Chocolate Sandwich Cookies        61             19     5.8   

   order_id  user_id  order_number  order_dow  order_hour_of_day  \
0   3139998      138            28          6                 11   
1   1977647      138            30          6                 17   
2    389851      709             2          0                 21   

   days_since_prior_order  First_order  add_to_cart_order  reordered _merge  \
0                     3.0        False                  5          0   both   
1                    20.0        False                  1          1   both   
2                     6.0        False                 20          0   both   

     Busiest_days Slowest_days busiest_period_of_day  max_order  \
0  Regularly busy                        Most orders         32   
1  Regularly busy                     Average orders         32   
2    Busiest days                      Fewest orders          5   

       loyalty_flag spending_flag order_frequency_flag  
0  Regular customer   Low spender    Frequent customer  
1  Regular customer   Low spender    Frequent customer  
2      New customer   Low spender    Frequent customer

In [23]:
## Dropping median days column as instructions dont specify it is needed.
ords_prods_merge.drop(columns=['median_days'], inplace=True)

In [25]:
## Export data 
ords_prods_merge.to_pickle(os.path.join(path,'02 - Data','prepared data','ords_prods_merge_task_4.8_Task.pkl'))